org.adempiere.base/src/org/compiere/util/Msg.java

```diff
@@ -143,30 +144,29 @@ private Msg()
 		ResultSet rs = null;
 		try
 		{
+			
 			if (AD_Language == null || AD_Language.length() == 0 || Env.isBaseLanguage(AD_Language, "AD_Language"))
 				pstmt = DB.prepareStatement("SELECT Value, MsgText, MsgTip FROM AD_Message",  null);
 			else
 			{
 				pstmt = DB.prepareStatement("SELECT m.Value, t.MsgText, t.MsgTip "
 					+ "FROM AD_Message_Trl t, AD_Message m "
 					+ "WHERE m.AD_Message_ID=t.AD_Message_ID"
+					+ " AND t.AD_Client_ID = 0" // load only translated messages at System level (using Value as key)
 					+ " AND t.AD_Language=?", null);
 				pstmt.setString(1, AD_Language);
 			}
-			rs = pstmt.executeQuery();
 
-			//	get values
-			while (rs.next())
-			{
-				String AD_Message = rs.getString(1);
-				StringBuilder MsgText = new StringBuilder();
-				MsgText.append(rs.getString(2));
-				String MsgTip = rs.getString(3);
-				//
-				if (MsgTip != null)			//	messageTip on next line, if exists
-					MsgText.append(" ").append(SEPARATOR).append(MsgTip);
-				msg.put(AD_Message, MsgText.toString());
-			}
+			addMessagesInCache(pstmt, rs, msg);
+
+			// load translated messages at tenant level (using AD_Client_ID|Value as key)
+			pstmt = DB.prepareStatement("SELECT t.AD_Client_ID || '|' || m.Value, t.MsgText, t.MsgTip"
```

Comment:
you need to close the previous pstmt before create a new one

Reply:
Anyway, it seems unnecessary to break the loading in 2 sql statement. You just need to retrieve AD_Client_ID and check that in the addMessagesInCache method (ie if (AD_Client_ID > 0) AD_Message = AD_Client_ID+"|"+AD_Message).

Reply:
In fact, yes we must => if you use Base language as language when you log in, initMsg will only load records from AD_Message (using the first query). The second one (which read data from both tables is used when user select a different language.

And I didn't find a solution to avoid those 2 separate queries

Reply:
ok but please the isactive and close of pstmt issue.

Reply:
Ok, should the rs be closed too ?
If so, DB.close(rs, pstmt); ?

Reply:
Just close pstmt is sufficient. rs is auto close when you close pstmt.

Reply:
ok, done